File: first-and-last-position-of-element.java
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


class Solution {
    public int[] searchRange(int[] nums, int target) {
        if(nums.length == 0) return new int[]{-1, -1};
        int[] index = new int[]{-1, -1};
        int left = 0, right = nums.length-1, mid = 0;
        while(left <= right){
            mid = (left + right)/2;
            if(target <= nums[mid]){
                right = mid-1;
            }else{
                left = mid+1;
            }
            if(target == nums[mid]){
                index[0] = mid;
            }
        }
        left = 0; right = nums.length-1; mid = 0;
         while(left <= right){
            mid = (left + right)/2;
            if(target < nums[mid]){
                right = mid-1;
            }else{
                left = mid+1;
            }
            if(target == nums[mid]){
                index[1] = mid;
            }
        }
        return index;
    }   
}

File: FirstAndLastPositionOfElement.py
# Problem 1: (https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if len(nums) == 0:
            return [-1,-1]
        low = 0
        high = len(nums) - 1
        first = -1
        while low <= high:   
            mid = low + (high - low)//2
            if nums[mid] == target:
                if mid != 0 and nums[mid-1] == target:
                    high = mid - 1
                else:
                    first = mid
                    break
            elif nums[mid] < target :
                low = mid + 1
            else:
                high = mid - 1
        last = -1
        low = first
        high = len(nums) - 1
        while low <= high:
            mid = low + (high - low)//2
            if nums[mid] == target:
                if mid != len(nums)-1 and nums[mid+1] == target:
                    low = mid + 1
                else:
                    last = mid
                    break
            elif nums[mid] < target :
                low = mid + 1
            else:
                high = mid - 1
        return [first, last]

#find the first occurence of element using binary search in range 0, n-1
#find the last occurence of element using binary search in range index of first occurence and n-1
#time complexity O(log2n)
#space complexity O(1)